protocol/0082-RFPR-on_chain_referral_program.md

```diff
@@ -0,0 +1,235 @@
+- all `referral_discount_factor` values must be less than or equal to the network parameter `referralProgram.maxReferralDiscountFactor`.
+- `window_length` must be an integer greater than zero.
+
+After a referral program has been passed, the referral program must be assigned an `id` by the network and exposed through an API. A program can then be updated through a governance proposal using this `id`.
+. 
+```protobuf
+message UpdateReferralProgram{
+    id: "abcd3fgh1jklmn0pqf5tuvwzyz"
+    changes: UpdateReferralProgramConfiguration{
+        benefit_tiers: [
+            {
+                "minimum_running_volume": 10000,
+                "referral_reward_factor": 0.001,
+                "referral_discount_factor": 0.002,
+            }
+        ],
+        start_epoch: 107
+        end_epoch: 114,
+        window_length: 1,
+    }
+}
+```
+In addition to the conditions which apply when creating a referral program, the following apply:
+- a proposer cannot update a program if its state is `STATUS_REJECTED`, `STATUS_STOPPED`, or `STATUS_CLOSED`.
+- a proposer cannot update the `name` of a program.
+- a proposer cannot update the `start_epoch` if the program is `STATUS_ACTIVE`
+
+## Referral Program Lifecycle:
+
+After a referral program [proposal](#governance-proposals) is validated and accepted by the network, a referral program is created and can be one of the following states. The current state of a program should be exposed via an API.
+
+| Status             | Benefits Enabled | Condition for entry                                       | Condition for exit                                   |
+| ------------------ | ---------------- | --------------------------------------------------------- | ---------------------------------------------------- |
+| `STATUS_PROPOSED`  | No               | Governance proposal valid and accepted                    | Governance proposal voting period ends               |
+| `STATUS_REJECTED`  | No               | Governance vote fails                                     | na                                                   |
+| `STATUS_PENDING`   | No               | Governance vote passes                                    | Current epoch greater than or equal to `start_epoch` |
+| `STATUS_STOPPED`   | No               | There is already an existing program with `STATUS_ACTIVE` | na                                                   |
+| `STATUS_ACTIVE`    | Yes              | Previously `STATUS_PENDING`                               | Current epoch less than or equal to `end_epoch`      |
+| `STATUS_CLOSED`    | No               | Previously `STATUS_ACTIVE`                                | na                                                   |              
+
+## Team Mechanics
+A team is comprised of a referrer and all their referees. There can only ever be one referrer per team but the number of referees is unlimited.
+
+### Creating a team
```

Comment:
We should make this "Creating or updating a team" and note that sending the transaction a second or subsequent time must not change the referal code/team ID but should cause the metadata (name, URL, image) to be updated.